synthesizer/process/src/lib.rs

```diff
@@ -84,6 +84,8 @@ pub struct Process<N: Network> {
     universal_srs: UniversalSRS<N>,
     /// The mapping of program IDs to stacks.
     stacks: Arc<RwLock<IndexMap<ProgramID<N>, Arc<Stack<N>>>>>,
+    /// The mapping of program IDs to old stacks.
+    old_stacks: Arc<RwLock<IndexMap<ProgramID<N>, Option<Arc<Stack<N>>>>>>,
```

Comment:
Open question on whether it's cleaner to stage stacks in `Process` or manage them out of band in `atomic_{finalize, speculate}`